[ONNX] Adding optional type in ONNX exporter

neginraoof opened the pull request:
1- Export optional type operators: Optional, OptionalHasElement, OptionalGetElement
2- Fix shape and type inference for controlflow ops (onnx::If) output shape/type inference for optional type (including optional uninitialized output)
3- Fix some minor issues in fold_if_node pass
4- Handle optional type block output
5- Enable onnx attribute of kind: "type" (TypeProto). This is needed for Optional constructor op to set element type. Added TypeProto graph encoder.
6- Fix input shape and type inference for optional type, resolve type from model forward function signature.
7- Add optional type serialization.
8- Enable export of ORT backend unit tests that can be supported with ONNX optional type (where tensor type is present)
9- Enable Model tests for models using Optional Type in test_models

neginraoof committed:
Add opset 14

neginraoof committed:
Update onnx build

neginraoof committed:
Fix for controlflow

neginraoof committed:
Fix for dynamic shape inference

neginraoof committed:
Fix controlflow

neginraoof committed:
Fix for controlflow

neginraoof committed:
Update opset 15

neginraoof committed:
Add _is_ op

neginraoof committed:
Merge

BowenBao commented on torch/csrc/jit/passes/onnx/fold_if_node.cpp:
```diff
@@ -123,7 +126,7 @@ static bool constantFoldedConditionValue(Node* node) {
   ScalarTypeAnalysisNodeForONNX(compare_node);
```
I wonder if this pass and file `fold_if_node.cpp` is still used? I think it may be covered completely with shape related folding updates by David. cc @jiafatom

jiafatom replied:
Currently `fold_if` pass is still in use. My WIP PR https://github.com/pytorch/pytorch/pull/62856 is going to deprecate this pass - I did some minor changes, then we can deprecate this `fold_if` pass.

jiafatom replied:
Update: Deprecate `fold_if` pass PR https://github.com/pytorch/pytorch/pull/65697 is merged into `onnx_ms_1`.

BowenBao commented on torch/csrc/jit/passes/onnx/peephole.cpp:
```diff
@@ -948,10 +948,10 @@ static void removeSequenceSplitConcat(Block* b) {
   for (auto out : b->outputs()) {
     auto n = out->node();
-    if (nullptr != n && n->kind() == prim::Param) {
+    if (out->node()->owningBlock() != b) {
```
nit
```suggestion
    if (n->owningBlock() != b) {
```

BowenBao replied:
curious if the following are always true?
1. out->node() can never be nullptr?
2. prim::Param node for a certain block is not owned by that block?

BowenBao commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -2602,15 +2602,16 @@ def is_floating_point(g, self):
-        return g.op("Constant", value_t=torch.BoolTensor([1]))
+            return g.op("Constant", value_t=torch.BoolTensor([1]))
+        return g.op("Constant", value_t=torch.BoolTensor([0]))
     return ne(g, self, other)
```
return eq?

neginraoof committed:
Fix for comments

neginraoof committed:
Initial work for Optional type export

neginraoof committed:
Merge

hwangdeyu commented on torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:
```diff
@@ -262,23 +252,64 @@ void InferShapeTypeForUninitializedOutput(
+    }
+    const_node->insertBefore(block->return_node());
+    uninitialized_output->replaceAllUsesWith(const_node->output());
+  } else if (auto output_type = other_output->type()->cast<ListType>()) {
```
Kindly remainder for rebasing latest code. Uninitialized Output List type was modified and added tests and ONNXShapeTypeInference in https://github.com/pytorch/pytorch/pull/60534.

hwangdeyu commented on torch/onnx/utils.py:
```diff
@@ -805,7 +817,7 @@ def set_names(node_list, name_list, descriptor):
     set_names(list(graph.inputs()), input_names, "input")
     set_names(list(graph.outputs()), output_names, "output")
 
-attr_pattern = re.compile("^(.+)_([ifstgz])$")
```
Just curious, what‘s the meaning of `z` at this regular expression? It never appeared in our PyTorch symbolic function files.

garymm committed:
[ONNX] Update onnxruntime to 1.9 for CI (#65029)

shubhambhokare1 committed:
[ONNX] Add dim argument to all symbolic (#66093)

* Add dim argument to all symbolic

* All symbolic depends on any symbolic

garymm committed:
[ONNX] Use Reciprocal operator instead of Div(1, x). (#65382)

* [ONNX] Use Reciprocal operator instead of Div(1, x).

This is a more readable and perhaps more performant way to export
torch.reciprocal.

* Use Reciprocal in caffe to operator to import onnx

Co-authored-by: take-cheeze <takechi101010@gmail.com>

BowenBao committed:
[ONNX] Support conv-bn fusion in blocks (#66152)

* Support conv-bn fusion in nested blocks

* avoid running script tests twice

neginraoof committed:
merge

neginraoof committed:
fix merge

neginraoof committed:
format

shubhambhokare1 committed:
[ONNX] Relax check on Prim::PythonOp nodes for ONNX_FALLTHROUGH (#66172)

* Relax check on Prim::PythonOp nodes for Onnx_fallthrough

* Add tests

BowenBao committed:
[ONNX] Export nn.Module call as ONNX local function (#63589)

* Add new argument to export api to enable users specifying `nn.Module` classes that they wish to be exported as local function in ONNX model.
* Refactor `torch/csrc/jit/serialization/export.cpp`, and remove redundant `EncoderBase` class.
* ~~Contains changes from #63268~~
* Depends on #63716 to update onnx submodule.

Co-authored-by: BowenBao <bowbao@microsoft.com>

ghstack-source-id: 92a95520cee1007c2b6d9c87698aba11b25147a1
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66140

BowenBao committed:
ONNX: Delete or document skipped ORT tests (#64470)

Delete test_list_remove. There's no point in testing conversion of
this model since TorchScript doesn't support it.

Add a link to an issue tracking test_embedding_bag_dynamic_input.

[ONNX] fix docs (#65379)

Mainly fix the sphinx build by inserting empty before
bulleted lists.

Also some minor improvements:
Remove superfluous descriptions of deprecated and ignored args.
The user doesn't need to know anything other than that they are
deprecated and ignored.

Fix custom_opsets description.

Make indentation of Raises section consistent with Args section.

[ONNX] publicize func for discovering unconvertible ops (#65285)

* [ONNX] Provide public function to discover all unconvertible ATen ops

This can be more productive than finding and fixing a single issue at a
time.

* [ONNX] Reorganize test_utility_funs

Move common functionality into a base class that doesn't define any
tests.

Add a new test for opset-independent tests. This lets us avoid running
the tests repeatedly for each opset.

Use simple inheritance rather than the `type()` built-in. It's more
readable.

* [ONNX] Use TestCase assertions rather than `assert`

This provides better error messages.

* [ONNX] Use double quotes consistently.

[ONNX] Fix code block formatting in doc (#65421)

ghstack-source-id: c3d9f14c596ee1f7c7af93fc52eccf4ba632f63c
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66143

BowenBao committed:
[ONNX] Update sum symbolic to handle dtypes (#64289)

* Update aten::sum symbolic for dtype

* Remove nesting and modify opeartor tests

* Fix expect files

[ONNX] Fix expect files added in #64289 (#65356)

ghstack-source-id: 0cfc431210bb69809aa7ca9d5e8f000a640434f4
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66141

BowenBao committed:
[ONNX] Add warning for inplace updates on tensor.shape in tracing mode (#63170)

* Add warning

* Lint and clang fixes

* Remove duplicate comments

* Added pitfalls section

* Modify sections

* Minor modifications

* Add underline to avoid doc build failures

ghstack-source-id: a7941e7a5c5facf5fdae0a114d5d92dcda9751c6
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66142

BowenBao committed:
[ONNX] Symbolic: dynamic input for OneHot, bool for Einsum (#65940)

Symbolic: dynamic input for OneHot, bool for Einsum

Co-authored-by: jiafatom <jiafa@microsoft.com>

ghstack-source-id: 4f70dd0a7a18d401568c883efbd8f8ec158c27d7
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66147

BowenBao committed:
[ONNX] Deprecate fold_if pass (#65697)

Deprecate fold_if pass

Co-authored-by: jiafatom <jiafa@microsoft.com>

ghstack-source-id: fb1293a9a6ce96981191616bf765591c96afcc43
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66145

BowenBao committed:
[ONNX] Modify softplus symbolic to support beta!=1 (#65001)

* Modify softplus symbolic to support beta!=1

* Remove parse args

ghstack-source-id: 2db3a2a045e52031e94382d5f9a570936c5166d5
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66146

BowenBao committed:
[ONNX] Add diagonal symbolic (#64454)

* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

ghstack-source-id: 8690ae4651af0aa1caa66fccd0be13f55eef4947
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66144

BowenBao committed:
[ONNX] Update slice process shape to support rank only inference (#65782)

Updated logic will be able to infer rank of slice output, when only rank is known for slice input. Enables cases where `ConstantValueMap::HasRank(input)` is `True`, while `ConstantValueMap::HasShape(input)` is `False`.

Co-authored-by: BowenBao <bowbao@microsoft.com>

ghstack-source-id: 01477ba2671657fc432bf817f0317e3d1809402d
Pull Request resolved: https://github.com/pytorch/pytorch/pull/66149

neginraoof committed:
Merge commit 'refs/pull/61938/head' of https://github.com/pytorch/pytorch into neraoof/fixInputSequence

fatcat-z committed:
[ONNX] Remove the argument enable_onnx_checker of export() method entirely. (#66611)

[ONNX] Remove argument _retain_param_name from torch.onnx.export() function.

Co-authored-by: fatcat-z <jiz@microsoft.com>

fatcat-z committed:
Deprecate the argument _retain_param_name of export() method entirely. (#66617)

Remove the argument _retain_param_name of export() method entirely.

Co-authored-by: fatcat-z <jiz@microsoft.com>

BowenBao committed:
[ONNX] Update onnx function export with comments and clean up (#66817)

* Addresses comments from #63589

Co-authored-by: BowenBao <bowbao@microsoft.com>

fatcat-z committed:
Remove the argument strip_doc_string of export() method entirely. (#66615)

Remove the argument strip_doc_string of export() method entirely.

Co-authored-by: fatcat-z <jiz@microsoft.com>

BowenBao committed:
[ONNX] Update value name copying logic for onnx (#66170)

Specifically targets the symbolic functions that directly returns input as output. The old logic will override the value name with output value name. But since the input is unmodified and unchanged, it is more logically to keep its original input name. Especially for cases where inputs are directly from model parameters.

Co-authored-by: BowenBao <bowbao@microsoft.com>

garymm committed:
[ONNX] remove torch::onnx::PRODUCER_VERSION (#67107)

Use constants from version.h instead.
This simplifies things since we no longer have to update
PRODUCER_VERSION for each release.

Also add TORCH_VERSION to version.h so that a string is available for
this purpose.

BowenBao committed:
[ONNX] Suppress ort warnings in onnx related test (#67054)

Improve readability of test logs by suppressing ort warnings logging for onnx related test.

Reducing ONNX CI test log binary size:
linux-xenial-py3.6-clang7-onnx-test1: 12443 KB -> 6958 KB
linux-xenial-py3.6-clang7-onnx-test2: 16884 KB -> 5778 KB

Co-authored-by: BowenBao<bowbao@microsoft.com>

garymm committed:
[ONNX] Support opset 15 (#67121)

Also fix Reduce ops on binary_cross_entropy_with_logits

The graph says the output is a scalar but with `keepdims=1`
(the default), the output should be a tensor of rank 1. We set keep
`keepdims=0` to make it clear that we want a scalar output.

This previously went unnoticed because ONNX Runtime does not strictly
enforce shape inference mismatches if the model is not using the latest
opset version.

Co-authored-by: Bowen Bao <bowbao@microsoft.com>

garymm committed:
[ONNX] Fix new_full and full_like for Python 3.9 (#67124)

Previously new_full would fail with errors like:
`TypeError: only integer tensors of a single element can be converted to an index`

And full_like would trigger warnings like:
`DeprecationWarning: an integer is required (got type float).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.`

Co-authored-by: Bowen Bao <bowbao@microsoft.com>

neginraoof committed:
Merge branch 'onnx_ms_1' into neraoof/optional

neginraoof committed:
Merge

garymm committed:
[ONNX] Use human readable enum for dtype scalars (#66822)

Also make quoting of string literals consistent.

garymm committed:
[ONNX] Fix reciprocal when input is not floating point (#67471)

torch.reciprocal implicitly casts the inputs to float, and ONNX
Reciprocal requires floating point inputs.

Also separate the reciprocal test from other tests, and test different
input types.

hwangdeyu committed:
[ONNX] Remove the argument example_outpus of export() method entirely. (#67082)

* remove the argument example_outpus of export() method entirely

Co-authored-by: hwangdeyu <dejack953@outlook.com>

Commit:
[ONNX] Allow registration of custom symbolics for aten namespace (#66481)

neginraoof committed:
Merge branch 'onnx_ms_1' into neraoof/optional

garymm commented:
I'm continuing this work in https://github.com/pytorch/pytorch/pull/68793, closing this.
